File: app/src/main/java/com/chooloo/www/callmanager/ui/recents/RecentsFragment.java
package com.chooloo.www.callmanager.ui.recents;

import android.database.Cursor;
import android.os.Bundle;

import androidx.annotation.Nullable;
import androidx.loader.content.Loader;

import com.chooloo.www.callmanager.cursorloader.RecentsCursorLoader;
import com.chooloo.www.callmanager.entity.RecentCall;
import com.chooloo.www.callmanager.ui.cursor.CursorFragment;

import static android.Manifest.permission.READ_CALL_LOG;
import static android.Manifest.permission.WRITE_CALL_LOG;

public class RecentsFragment extends CursorFragment<RecentsAdapter> implements RecentsMvpView {

    private static final String[] REQUIRED_PERMISSIONS = {READ_CALL_LOG, WRITE_CALL_LOG};
    private static final String ARG_CONTACT_NAME = "contact_name";
    private static final String ARG_PHONE_NUMBER = "phone_number";

    private RecentsPresenter<RecentsMvpView> mPresenter;

    public static RecentsFragment newInstance() {
        return RecentsFragment.newInstance(null, null);
    }

    public static RecentsFragment newInstance(@Nullable String phoneNumber, @Nullable String contactName) {
        Bundle args = new Bundle();
        args.putString(ARG_PHONE_NUMBER, phoneNumber);
        args.putString(ARG_CONTACT_NAME, contactName);
        RecentsFragment fragment = new RecentsFragment();
        fragment.setArguments(args);
        return fragment;
    }

    @Override
    public RecentsAdapter onGetAdapter() {
        RecentsAdapter recentsAdapter = new RecentsAdapter(activity);
        recentsAdapter.setOnRecentItemClickListener(recentCall -> mPresenter.onRecentItemClick(recentCall));
        recentsAdapter.setOnRecentItemLongClickListener(recentCall -> mPresenter.onRecentItemLongClick(recentCall));
        return recentsAdapter;
    }

    @Override
    public String[] onGetPermissions() {
        return REQUIRED_PERMISSIONS;
    }

    @Override
    public Loader<Cursor> onGetLoader(Bundle args) {
        String contactName = args.getString(ARG_CONTACT_NAME, null);
        String phoneNumber = args.getString(ARG_PHONE_NUMBER, null);
        return new RecentsCursorLoader(activity, phoneNumber, contactName);
    }

    @Override
    public void onSetup() {
        super.onSetup();

        mLoaderId = 1;

        mPresenter = new RecentsPresenter<>();
        mPresenter.attach(this);

        load();
    }

    @Override
    public void onDestroyView() {
        super.onDestroyView();
        mPresenter.detach();
    }

    @Override
    public void load(@Nullable String phoneNumber, @Nullable String contactName) {
        Bundle args = new Bundle();
        args.putString(ARG_PHONE_NUMBER, phoneNumber);
        args.putString(ARG_CONTACT_NAME, contactName);
        setArguments(args);
        load();
    }

    @Override
    public void openRecent(RecentCall recentCall) {
        // TODO implement
    }
}

File: app/src/main/java/com/chooloo/www/callmanager/ui/recents/RecentsFragment.kt
package com.chooloo.www.callmanager.ui.recents

import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModelProvider
import com.chooloo.www.callmanager.adapter.RecentsAdapter
import com.chooloo.www.callmanager.entity.Recent
import com.chooloo.www.callmanager.livedata.RecentsLiveData
import com.chooloo.www.callmanager.ui.list.ListFragment
import com.chooloo.www.callmanager.viewmodel.data.DataViewModel
import com.chooloo.www.callmanager.viewmodel.data.DataViewModelFactory

class RecentsFragment : ListFragment<RecentsAdapter>(), RecentsMvpView {
    private lateinit var _presenter: RecentsPresenter<RecentsMvpView>
    private lateinit var _recentsLiveData: RecentsLiveData

    companion object {
        @JvmStatic
        fun newInstance(): RecentsFragment {
            return RecentsFragment()
        }
    }

    override fun onGetAdapter(): RecentsAdapter {
        return RecentsAdapter(_activity).apply {
            setOnRecentItemClickListener { recent -> _presenter.onRecentItemClick(recent) }
            setOnRecentItemLongClickListener { recent -> _presenter.onRecentItemLongClick(recent) }
        }
    }

    override fun onSetup() {
        super.onSetup()

        _presenter = RecentsPresenter()
        _presenter.attach(this)

        _recentsLiveData = ViewModelProvider(this, DataViewModelFactory(_activity)).get(DataViewModel::class.java).recents

        try {
            _recentsLiveData.observe(viewLifecycleOwner, Observer { recents -> adapter.updateRecents(recents) })
        } catch (e: SecurityException) {
            showNoPermissions(true)
        }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        _presenter.detach()
    }

    override fun openRecent(recent: Recent) {
        // TODO implement
    }
}